### ISSUE TITLE
ordering issues with atomic parser

### ISSUE THREAD

**Issue by modosc** [acss-io/atomizer#260]
we had some styling issues with our site when using [atomic-plugin] (https://github.com/mdrohmann/atomic-plugin) but the site worked fine when using `atomizer` on the command line. i narrowed it down to the order in which files were being parsed - inside of `webpack` files were concatenated together based on the module require tree rather than the glob we were using on the command line.

example:

```
> var test1 = '<div class="D(n)"></div><div class="D(ib) D(n)">test</div>'
> var test2 = '<div class="D(ib) D(n)">test</div><div class="D(n)"></div>'

> console.log(atomizer.getCss(atomizer.getConfig(atomizer.findClassNames(test1), {})))
.D\(n\) {
  display: none;
}
.D\(ib\) {
  display: inline-block;
}

> console.log(atomizer.getCss(atomizer.getConfig(atomizer.findClassNames(test2), {})))
.D\(ib\) {
  display: inline-block;
}
.D\(n\) {
  display: none;
}
```

`test1` will be rendered with `display: inline-block`, `test2` with `display: none`.

we wound up calling `sort()` on the classes that we pass to `atomizer.getConfig` and that resolved the issue by making sure that `.D(n)` always had priority over `.D(i)`, `.D(ib)`, etc but that's just dumb luck that `none` sorts last. i think there should be some explicit ordering of these classes when they're generated so that the same files always render the same, regardless of the order of parsing.

**Comment by modosc** [comment]
another example is with media queries - using this config:
```
  breakPoints: {
    sm: '@media screen and (max-width: 900px)'
  }
```
these two blocks display differently:
```
<div class="W(1/3) W(100%)--sm">
```

```
<div class=" W(100%)--sm W(1/3)">
```

keep in mind it's not just the ordering of the classes here - if a different file references one or other other class _before_ this one it will change the order of the generated css and thus the behavior. 

also the second example's css looks like this:
```
@media screen and (max-width: 900px) {
  .W\(100\%\)--sm {
    width: 100%;
  }
}
  .W\(1\/3\) {
    width: 33.3333%;
  }
```

(note the indenting after the `@media` section)

**Comment by dnewcome** [comment]
We had a discussion about this just now. I think the intent with the acss class names is that there is no implied specificity based on order. It's not expected that `D(n)` will necessarily override a previous `D(ib)` on the same element. However, we do expect a breakpoint `D(n)--sm` to override for example. The real issue is with media query ordering since `@media` doesn't provide any additional specificity and is dependent on appearing later in the cascade.

**Comment by thierryk** [comment]
The idea behind ACSS is to rely on _flat selectors_ (selectors of same specificity&mdash;`0,0,1,0`). 
As @dnewcome explained, declarations meant to **override** "basic" classes (i.e. "breakpoint" classes) must appear last to leverage the cascade.

So I think the question is more about the use of classes (ACSS or else) like this:

```
<div class="D(ib) D(n)">test</div>
```

What is the expectation here? We can't set both styles so which one should "win", and **why**?

Note that if `D(n)` is applied via JavaScript to overwrite `D(ib)` then it is better to use `D(n)!` for that as it is mapped to: `display: none !important`. That, or simply replace `D(ib)` with `D(n)`.

**Comment by modosc** [comment]
i agree that the first example feels sort of contrived - it was in prod code but we can obviously work around it if need be. i think that it's indicative of a real issue though.

if i'm understanding you correctly about having breakpoint classes appearing last i can still cause breakage easily:
```
<div class="W(100%)--sm"></div>
<div class="W(1/3) W(100%)--sm">Test</div>
```
vs
```
<div class="W(1/3)"></div>
<div class="W(1/3) W(100%)--sm">Test</div>
```

just having one or the other class appear first in the document (or any other document) is enough to give it precedence, and if you're parsing an entire tree of documents this is entirely possible. 

or is the idea that if i'm _ever_ going to use a breakpoint class i have guarantee i've first used a non-breakpoint version? if so why not implement this in the compiler (or at least give a warning back)? granted you're now generating css for classes that aren't ever used but you'd fix issues like this.

**Comment by thierryk** [comment]
@modosc I think you're in the wrong repo. This is not a problem with Atomizer but with the [atomic plugin](https://github.com/mdrohmann/atomic-plugin). As I said earlier, it is really important that some rules (i.e. "breakpoint" ones) appear last in the sheet and Atomizer does this properly as you can verify yourself using [Pankaj's tool](https://pankajparashar.com/atomizer-web/).

Make sense?

**Comment by modosc** [comment]
@thierryk please check my initial examples again - they are using the `atomizer` api directly so the issue has nothing to do with `atomic-plugin`. [here's](https://github.com/modosc/atomizer-test-case) a repo to show the issue - in both `broken` and `works` there are 2 identical html files, `a.html` and `b.html`. to build the css i used `npm run build`:
```
rm -f works/atomic.css; 
rm -f broken/atomic.css;
./node_modules/.bin/atomizer -c config.js works/a.html works/b.html > works/atomic.css
./node_modules/.bin/atomizer -c config.js broken/b.html broken/a.html > broken/atomic.css
```

note the only difference is the ordering of the input files.

now open `a.html` from both `broken` and `works` and resize your browser's width to `< 900px`. 

in `works` the image picks up the breakpoint rule and changes width to `100%` but in `broken` the image stays at `33%`. this is due to the ordering of the generated css which in turn comes from the order these files were processed in. that's the bug i see here - i should get the same css output when i process a given set of input files but i don't when the ordering of the input files changes.

**Comment by thierryk** [comment]
@modosc Yikes! You're right. I tried a simple `atomizer -o atomic.css *.html` _after_ inverting the content of the 2 html files and I obtained the same result: the rule related to media queries _does not come last_ (**after** basic ACSS rules).

Thank you for reporting this and sorry for not catching the issue earlier in this thread.

Pinging @renatoi and @src-code

**Comment by renatoi** [comment]
I'll take a look at that probably next week, can't do it right now, I'd appreciate a PR if someone has time.

**Comment by src-code** [comment]
@modosc @thierryk @renatoi Have a fix in PR #265, let me know what you think.
